feat(Constraint): Component constraint authoring

geoffreykwan opened the pull request:
## Changes

- Implemented component constraint authoring component
- Added constraint authoring component to all the component advanced authoring popups
- Created domain objects ValueAndText, RemovalCriteria, RemovalCriteriaParam

## Test

- Make sure component constraint authoring works for all the components
- Make sure the step constraint authoring still works like before

Closes #1137

geoffreykwan committed:
feat(Constraint): Component constraint authoring

#1137

hirokiterashima commented on src/assets/wise5/authoringTool/constraint/constraint-authoring/constraint-authoring.component.ts:
```diff
@@ -1,147 +1,91 @@
+  REMOVAL_CRITERIA_STEP_PARAM: RemovalCriteriaParam = new RemovalCriteriaParam(
+    'nodeId',
+    $localize`Step`
+  );
```
Do we need to uppercase these variables? These are local variables, so I think we should make them lowercase and camelCase. Uppercase and underscores make the name longer and harder to read.

hirokiterashima commented on src/assets/wise5/authoringTool/constraint/node-constraint-authoring/node-constraint-authoring.component.ts:
```diff
@@ -0,0 +1,38 @@
+    ),
+    new ValueAndText('makeAllOtherNodesNotVisible', $localize`Make all other nodes not visible`),
+    new ValueAndText('makeThisNodeNotVisitable', $localize`Make this node not visitable`),
+    new ValueAndText('makeThisNodeNotVisible', $localize`Make this node not visible`)
```
Should we define a ConstraintAction class? I think it'd be more clear than ValueAndText.

hirokiterashima commented on src/app/domain/removalCriteriaParam.ts:
```diff
@@ -0,0 +1,12 @@
+export class RemovalCriteriaParam extends ValueAndText {
+  defaultValue: any;
+  text: string;
+  value: string;
```
Are these redundant because they're declared in the parent?

hirokiterashima commented on src/assets/wise5/authoringTool/constraint/component-constraint-authoring/component-constraint-authoring.component.ts:
```diff
@@ -0,0 +1,24 @@
+export class ComponentConstraintAuthoringComponent extends ConstraintAuthoringComponent {
+  constraintActions = [
+    new ValueAndText('', $localize`Please Choose an Action`),
+    new ValueAndText('makeThisComponentNotVisible', $localize`Make This Component Not Visible`)
```
Create ConstraintAction class?

hirokiterashima commented on src/app/authoring-tool/edit-component-constraints/edit-component-constraints.component.ts:
```diff
@@ -0,0 +1,62 @@
+  templateUrl: './edit-component-constraints.component.html',
+  styleUrls: ['./edit-component-constraints.component.scss']
+})
+export class EditComponentConstraintsComponent implements OnInit {
```
This class has a lot of duplicated code with NodeAdvancedConstraintAuthoringComponent. Can we refactor to reduce duplication?

hirokiterashima reviewed (changes requested):
Functionality works well. I added some code comments inline, some are suggested changes and some may need further discussion.

Some UI/UX-related comments:
- Right now you can add two "Make this Component not Visible" constraints on a component. Should we prevent this?
- Should we choose "Make this Component not Visible" option by default since that's the only selectable option right now?
- Component constraint authoring: should we show a border around it to delineate from other component authoring options?
- Choice Chosen constraint removal criteria authoring should only show MC steps and components as available options? Right now it shows all steps and components. This may be better handled in a separate task.

hirokiterashima commented on src/assets/wise5/authoringTool/constraint/node-constraint-authoring/node-constraint-authoring.component.ts:
```diff
@@ -0,0 +1,38 @@
+    new ValueAndText('makeThisNodeNotVisible', $localize`Make this node not visible`)
+  ];
+
+  constructor(
+    protected dataService: TeacherDataService,
+    protected projectService: TeacherProjectService
+  ) {
+    super(dataService, projectService);
+  }
```
Not needed?
```suggestion
```

hirokiterashima commented on src/assets/wise5/authoringTool/constraint/component-constraint-authoring/component-constraint-authoring.component.ts:
```diff
@@ -0,0 +1,24 @@
+    protected projectService: TeacherProjectService
+  ) {
+    super(dataService, projectService);
+  }
```
Not needed?
```suggestion
```

geoffreykwan committed:
refactor(Constraint): Clean up component constraint authoring code

#1137

geoffreykwan committed:
refactor(Constraint): Create parent ConstraintsAuthoringComponent

Also only allow adding one component constraint and automatically set
the constraint action if there is only one possible action.

#1137

hirokiterashima reviewed (approved):
Looks good.

In the component constraint authoring, can we remove the "Please Choose an Action" option? I don't think it makes sense to let them choose this option right now, when we only have one option. Then, will we still need ```automaticallySetActionIfPossible()```, or can we just always choose the first constraintAction in the list?
